File: frontend/src/core/AutomationLane.ts
/**
 * Automation Lane System
 *
 * DAW-style automation lanes for precise parameter control over time.
 * Supports multiple interpolation modes, recording, and real-time playback.
 *
 * Features:
 * - Timeline-synchronized automation points
 * - Multiple interpolation modes (linear, bezier, step, exponential)
 * - Recording mode for real-time automation capture
 * - Touch/Latch/Write recording modes
 * - Parameter snapping and quantization
 * - Undo/redo support via command pattern
 *
 * Architecture:
 * ```
 * [AutomationLane] ── manages ── [AutomationPoint[]]
 *       │
 *       ├── getValue(time) → interpolated value
 *       ├── record(time, value) → adds/updates points
 *       └── connect(target) → applies to parameter
 * ```
 *
 * @module core/AutomationLane
 */

import { rfDebug } from './dspMetrics';

// ============ Types ============

export type InterpolationMode = 'linear' | 'bezier' | 'step' | 'exponential' | 'smooth';

export type RecordingMode = 'off' | 'touch' | 'latch' | 'write';

export interface AutomationPoint {
  /** Time in seconds from start */
  time: number;
  /** Parameter value (normalized 0-1 or raw depending on target) */
  value: number;
  /** Interpolation to next point */
  interpolation: InterpolationMode;
  /** Bezier control points [x1, y1, x2, y2] for cubic bezier */
  bezierHandles?: [number, number, number, number];
  /** Whether this point is selected (for UI) */
  selected?: boolean;
}

export interface AutomationLaneConfig {
  /** Unique ID */
  id: string;
  /** Display name */
  name: string;
  /** Target parameter path (e.g., "master.inserts.0.params.gain") */
  targetPath: string;
  /** Parameter range */
  range: {
    min: number;
    max: number;
    /** Default value */
    default: number;
    /** Value step/resolution */
    step?: number;
  };
  /** Display color (hex) */
  color?: string;
  /** Whether lane is visible in UI */
  visible?: boolean;
  /** Whether lane is locked (no editing) */
  locked?: boolean;
  /** Default interpolation for new points */
  defaultInterpolation?: InterpolationMode;
}

export interface AutomationLaneState {
  config: AutomationLaneConfig;
  points: AutomationPoint[];
  /** Current recording mode */
  recordingMode: RecordingMode;
  /** Is currently recording */
  isRecording: boolean;
  /** Last recorded time (for touch mode) */
  lastRecordTime: number;
  /** Last read time (for interpolation cache) */
  lastReadTime: number;
  /** Cached interpolated value */
  cachedValue: number;
}

// ============ Interpolation Functions ============

/**
 * Linear interpolation between two values.
 */
function lerp(a: number, b: number, t: number): number {
  return a + (b - a) * t;
}

/**
 * Cubic bezier interpolation.
 */
function cubicBezier(
  t: number,
  p0: number,
  p1: number,
  p2: number,
  p3: number
): number {
  const oneMinusT = 1 - t;
  return (
    oneMinusT * oneMinusT * oneMinusT * p0 +
    3 * oneMinusT * oneMinusT * t * p1 +
    3 * oneMinusT * t * t * p2 +
    t * t * t * p3
  );
}

/**
 * Smooth step interpolation (ease in-out).
 */
function smoothstep(t: number): number {
  return t * t * (3 - 2 * t);
}

/**
 * Exponential interpolation.
 */
function exponential(a: number, b: number, t: number, curve: number = 2): number {
  // curve > 1 = ease out, curve < 1 = ease in
  const tCurved = Math.pow(t, curve);
  return lerp(a, b, tCurved);
}

// ============ Automation Lane Class ============

/**
 * Manages automation data for a single parameter.
 */
export class AutomationLane {
  private state: AutomationLaneState;
  private updateCallback: ((value: number) => void) | null = null;
  private recording = false;
  private recordBuffer: AutomationPoint[] = [];
  private recordStartTime = 0;

  constructor(config: AutomationLaneConfig) {
    this.state = {
      config,
      points: [],
      recordingMode: 'off',
      isRecording: false,
      lastRecordTime: 0,
      lastReadTime: -1,
      cachedValue: config.range.default,
    };
  }

  // ============ Point Management ============

  /**
   * Add an automation point.
   */
  addPoint(point: Omit<AutomationPoint, 'interpolation'> & { interpolation?: InterpolationMode }): void {
    if (this.state.config.locked) return;

    const newPoint: AutomationPoint = {
      ...point,
      interpolation: point.interpolation ?? this.state.config.defaultInterpolation ?? 'linear',
    };

    // Insert in sorted order by time
    const insertIndex = this.findInsertIndex(newPoint.time);
    this.state.points.splice(insertIndex, 0, newPoint);

    rfDebug('AutomationLane', `Added point at ${newPoint.time.toFixed(3)}s: ${newPoint.value}`);
  }

  /**
   * Remove an automation point by index.
   */
  removePoint(index: number): void {
    if (this.state.config.locked) return;
    if (index >= 0 && index < this.state.points.length) {
      this.state.points.splice(index, 1);
    }
  }

  /**
   * Update an automation point.
   */
  updatePoint(index: number, updates: Partial<AutomationPoint>): void {
    if (this.state.config.locked) return;
    if (index >= 0 && index < this.state.points.length) {
      this.state.points[index] = { ...this.state.points[index], ...updates };
      // Re-sort if time changed
      if (updates.time !== undefined) {
        this.state.points.sort((a, b) => a.time - b.time);
      }
    }
  }

  /**
   * Move a point to a new time/value.
   */
  movePoint(index: number, time: number, value: number): void {
    this.updatePoint(index, { time, value });
  }

  /**
   * Clear all automation points.
   */
  clearPoints(): void {
    if (this.state.config.locked) return;
    this.state.points = [];
  }

  /**
   * Set all points at once (for undo/redo).
   */
  setPoints(points: AutomationPoint[]): void {
    if (this.state.config.locked) return;
    this.state.points = [...points].sort((a, b) => a.time - b.time);
  }

  /**
   * Get all points.
   */
  getPoints(): AutomationPoint[] {
    return [...this.state.points];
  }

  // ============ Value Interpolation ============

  /**
   * Get interpolated value at a specific time.
   */
  getValue(time: number): number {
    // Use cache if time hasn't changed
    if (time === this.state.lastReadTime) {
      return this.state.cachedValue;
    }

    const points = this.state.points;
    const config = this.state.config;

    // No points - return default
    if (points.length === 0) {
      return config.range.default;
    }

    // Before first point
    if (time <= points[0].time) {
      return points[0].value;
    }

    // After last point
    if (time >= points[points.length - 1].time) {
      return points[points.length - 1].value;
    }

    // Find surrounding points
    let p0 = points[0];
    let p1 = points[1];

    for (let i = 0; i < points.length - 1; i++) {
      if (time >= points[i].time && time < points[i + 1].time) {
        p0 = points[i];
        p1 = points[i + 1];
        break;
      }
    }

    // Calculate interpolation factor
    const t = (time - p0.time) / (p1.time - p0.time);

    // Interpolate based on mode
    let value: number;

    switch (p0.interpolation) {
      case 'step':
        value = p0.value;
        break;

      case 'linear':
        value = lerp(p0.value, p1.value, t);
        break;

      case 'smooth':
        value = lerp(p0.value, p1.value, smoothstep(t));
        break;

      case 'exponential':
        value = exponential(p0.value, p1.value, t);
        break;

      case 'bezier':
        if (p0.bezierHandles) {
          // Use bezier handles for value curve
          const [, y1, , y2] = p0.bezierHandles;
          // Approximate bezier by using y values as control points
          value = cubicBezier(t, p0.value, p0.value + y1 * (p1.value - p0.value),
            p1.value + y2 * (p0.value - p1.value), p1.value);
        } else {
          value = lerp(p0.value, p1.value, smoothstep(t));
        }
        break;

      default:
        value = lerp(p0.value, p1.value, t);
    }

    // Clamp to range
    value = Math.max(config.range.min, Math.min(config.range.max, value));

    // Apply step quantization if defined
    if (config.range.step) {
      value = Math.round(value / config.range.step) * config.range.step;
    }

    // Cache result
    this.state.lastReadTime = time;
    this.state.cachedValue = value;

    return value;
  }

  /**
   * Get value and apply to connected callback.
   */
  tick(time: number): number {
    const value = this.getValue(time);
    if (this.updateCallback) {
      this.updateCallback(value);
    }
    return value;
  }

  // ============ Recording ============

  /**
   * Set recording mode.
   */
  setRecordingMode(mode: RecordingMode): void {
    this.state.recordingMode = mode;
    if (mode === 'off') {
      this.stopRecording();
    }
  }

  /**
   * Start recording automation.
   */
  startRecording(startTime: number): void {
    if (this.state.config.locked) return;
    if (this.state.recordingMode === 'off') return;

    this.recording = true;
    this.state.isRecording = true;
    this.recordStartTime = startTime;
    this.recordBuffer = [];

    // In write mode, clear existing points from start time
    if (this.state.recordingMode === 'write') {
      this.state.points = this.state.points.filter(p => p.time < startTime);
    }

    rfDebug('AutomationLane', `Started recording (${this.state.recordingMode})`);
  }

  /**
   * Record a value at a specific time.
   */
  record(time: number, value: number): void {
    if (!this.recording) return;

    const point: AutomationPoint = {
      time,
      value: Math.max(this.state.config.range.min,
        Math.min(this.state.config.range.max, value)),
      interpolation: this.state.config.defaultInterpolation ?? 'linear',
    };

    // Add to buffer for later processing
    this.recordBuffer.push(point);
    this.state.lastRecordTime = time;
  }

  /**
   * Stop recording and finalize points.
   */
  stopRecording(): void {
    if (!this.recording) return;

    this.recording = false;
    this.state.isRecording = false;

    // Process record buffer - thin out redundant points
    const thinnedPoints = this.thinRecordBuffer(this.recordBuffer);

    // Merge with existing points based on mode
    if (this.state.recordingMode === 'write') {
      // Already cleared, just add new points
      for (const point of thinnedPoints) {
        this.addPoint(point);
      }
    } else {
      // Touch/Latch: replace points in recorded range
      if (thinnedPoints.length > 0) {
        const startTime = thinnedPoints[0].time;
        const endTime = thinnedPoints[thinnedPoints.length - 1].time;

        // Remove existing points in range
        this.state.points = this.state.points.filter(
          p => p.time < startTime || p.time > endTime
        );

        // Add new points
        for (const point of thinnedPoints) {
          this.addPoint(point);
        }
      }
    }

    this.recordBuffer = [];
    rfDebug('AutomationLane', `Stopped recording, added ${thinnedPoints.length} points`);
  }

  /**
   * Thin out redundant points from recording buffer.
   * Keeps points where value changed significantly.
   */
  private thinRecordBuffer(points: AutomationPoint[]): AutomationPoint[] {
    if (points.length <= 2) return points;

    const threshold = (this.state.config.range.max - this.state.config.range.min) * 0.01;
    const thinned: AutomationPoint[] = [points[0]];

    for (let i = 1; i < points.length - 1; i++) {
      const prev = thinned[thinned.length - 1];
      const curr = points[i];
      const next = points[i + 1];

      // Keep point if value differs significantly from linear interpolation
      const expectedValue = lerp(prev.value, next.value,
        (curr.time - prev.time) / (next.time - prev.time));
      const diff = Math.abs(curr.value - expectedValue);

      if (diff > threshold) {
        thinned.push(curr);
      }
    }

    thinned.push(points[points.length - 1]);
    return thinned;
  }

  // ============ Target Connection ============

  /**
   * Connect to a parameter update callback.
   */
  connect(callback: (value: number) => void): void {
    this.updateCallback = callback;
  }

  /**
   * Disconnect from parameter.
   */
  disconnect(): void {
    this.updateCallback = null;
  }

  // ============ Utility ============

  /**
   * Find insertion index for sorted array.
   */
  private findInsertIndex(time: number): number {
    let low = 0;
    let high = this.state.points.length;

    while (low < high) {
      const mid = Math.floor((low + high) / 2);
      if (this.state.points[mid].time < time) {
        low = mid + 1;
      } else {
        high = mid;
      }
    }

    return low;
  }

  /**
   * Get configuration.
   */
  getConfig(): AutomationLaneConfig {
    return this.state.config;
  }

  /**
   * Update configuration.
   */
  updateConfig(updates: Partial<AutomationLaneConfig>): void {
    this.state.config = { ...this.state.config, ...updates };
  }

  /**
   * Get current state (for serialization).
   */
  getState(): AutomationLaneState {
    return { ...this.state };
  }

  /**
   * Check if lane is recording.
   */
  isRecording(): boolean {
    return this.recording;
  }

  /**
   * Get recording start time.
   */
  getRecordingStartTime(): number {
    return this.recordStartTime;
  }

  /**
   * Get point at or near a time (for selection).
   */
  getPointNear(time: number, tolerance: number): { point: AutomationPoint; index: number } | null {
    for (let i = 0; i < this.state.points.length; i++) {
      if (Math.abs(this.state.points[i].time - time) <= tolerance) {
        return { point: this.state.points[i], index: i };
      }
    }
    return null;
  }

  /**
   * Select points in a time range.
   */
  selectPointsInRange(startTime: number, endTime: number): void {
    for (const point of this.state.points) {
      point.selected = point.time >= startTime && point.time <= endTime;
    }
  }

  /**
   * Clear all selections.
   */
  clearSelection(): void {
    for (const point of this.state.points) {
      point.selected = false;
    }
  }

  /**
   * Delete selected points.
   */
  deleteSelected(): void {
    if (this.state.config.locked) return;
    this.state.points = this.state.points.filter(p => !p.selected);
  }
}

// ============ Automation Manager ============

/**
 * Manages multiple automation lanes and coordinates playback.
 */
class AutomationManagerClass {
  private lanes: Map<string, AutomationLane> = new Map();
  private isPlaying = false;
  private currentTime = 0;
  private animationFrame: number | null = null;

  /**
   * Create a new automation lane.
   */
  createLane(config: AutomationLaneConfig): AutomationLane {
    const lane = new AutomationLane(config);
    this.lanes.set(config.id, lane);
    rfDebug('AutomationManager', `Created lane: ${config.id}`);
    return lane;
  }

  /**
   * Get an automation lane by ID.
   */
  getLane(id: string): AutomationLane | undefined {
    return this.lanes.get(id);
  }

  /**
   * Remove an automation lane.
   */
  removeLane(id: string): void {
    const lane = this.lanes.get(id);
    if (lane) {
      lane.disconnect();
      this.lanes.delete(id);
      rfDebug('AutomationManager', `Removed lane: ${id}`);
    }
  }

  /**
   * Get all lane IDs.
   */
  getLaneIds(): string[] {
    return Array.from(this.lanes.keys());
  }

  /**
   * Get all lanes.
   */
  getAllLanes(): AutomationLane[] {
    return Array.from(this.lanes.values());
  }

  /**
   * Start playback at a specific time.
   */
  play(startTime: number = 0): void {
    this.currentTime = startTime;
    this.isPlaying = true;
    this.tick();
    rfDebug('AutomationManager', `Started playback at ${startTime}`);
  }

  /**
   * Stop playback.
   */
  stop(): void {
    this.isPlaying = false;
    if (this.animationFrame !== null) {
      cancelAnimationFrame(this.animationFrame);
      this.animationFrame = null;
    }
    rfDebug('AutomationManager', 'Stopped playback');
  }

  /**
   * Seek to a specific time.
   */
  seek(time: number): void {
    this.currentTime = time;
    // Update all lanes to new time
    for (const lane of this.lanes.values()) {
      lane.tick(time);
    }
  }

  /**
   * Get values for all lanes at a specific time.
   */
  getValuesAt(time: number): Map<string, number> {
    const values = new Map<string, number>();
    for (const [id, lane] of this.lanes) {
      values.set(id, lane.getValue(time));
    }
    return values;
  }

  /**
   * Start recording on specified lanes.
   */
  startRecording(laneIds: string[], startTime: number): void {
    for (const id of laneIds) {
      const lane = this.lanes.get(id);
      if (lane) {
        lane.startRecording(startTime);
      }
    }
  }

  /**
   * Stop recording on all lanes.
   */
  stopRecording(): void {
    for (const lane of this.lanes.values()) {
      lane.stopRecording();
    }
  }

  /**
   * Record a value to a lane.
   */
  recordValue(laneId: string, time: number, value: number): void {
    const lane = this.lanes.get(laneId);
    if (lane) {
      lane.record(time, value);
    }
  }

  /**
   * Internal tick for playback.
   */
  private tick(): void {
    if (!this.isPlaying) return;

    // Update all lanes
    for (const lane of this.lanes.values()) {
      lane.tick(this.currentTime);
    }

    // Schedule next tick
    this.animationFrame = requestAnimationFrame(() => this.tick());
  }

  /**
   * Update current time (called by transport).
   */
  setCurrentTime(time: number): void {
    this.currentTime = time;
  }

  /**
   * Dispose all lanes.
   */
  dispose(): void {
    this.stop();
    for (const lane of this.lanes.values()) {
      lane.disconnect();
    }
    this.lanes.clear();
    rfDebug('AutomationManager', 'Disposed');
  }
}

// ============ Singleton Export ============

/**
 * Global automation manager.
 */
export const automationManager = new AutomationManagerClass();

// Cleanup on page unload
if (typeof window !== 'undefined') {
  window.addEventListener('beforeunload', () => {
    automationManager.dispose();
  });
}

export default automationManager;

File: flutter_ui/lib/models/timeline/automation_lane.dart
// Automation Lane Model — Timeline Parameter Automation
//
// Represents automation curves for volume, pan, RTPC parameters.
// Supports bezier interpolation, multiple curve types.

import 'dart:math' as math;

/// Interpolation curve type
enum CurveType {
  linear,      // Straight line
  bezier,      // Smooth curve
  step,        // Instant jump
  exponential, // Exponential curve
  logarithmic, // Logarithmic curve
}

/// Single automation point
class AutomationPoint {
  final String id;
  double time;              // Time in seconds
  double value;             // Normalized 0.0-1.0 (converted based on parameter)
  CurveType interpolation;

  AutomationPoint({
    required this.id,
    required this.time,
    required this.value,
    this.interpolation = CurveType.linear,
  });

  AutomationPoint copyWith({
    String? id,
    double? time,
    double? value,
    CurveType? interpolation,
  }) {
    return AutomationPoint(
      id: id ?? this.id,
      time: time ?? this.time,
      value: value ?? this.value,
      interpolation: interpolation ?? this.interpolation,
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'time': time,
    'value': value,
    'interpolation': interpolation.name,
  };

  factory AutomationPoint.fromJson(Map<String, dynamic> json) {
    return AutomationPoint(
      id: json['id'] as String,
      time: (json['time'] as num).toDouble(),
      value: (json['value'] as num).toDouble(),
      interpolation: CurveType.values.firstWhere(
        (t) => t.name == json['interpolation'],
        orElse: () => CurveType.linear,
      ),
    );
  }
}

/// Parameter type for automation
enum AutomationParameterType {
  volume,   // 0.0-2.0 (−∞ to +6dB)
  pan,      // −1.0 to +1.0 (L/R)
  rtpc,     // Custom range per RTPC
  trigger,  // Boolean on/off
}

/// Automation lane for a single parameter
class AutomationLane {
  final String id;
  final String parameterId;           // 'volume', 'pan', 'rtpc_winAmount'
  final AutomationParameterType type;
  final List<AutomationPoint> points;
  final Color curveColor;
  final double minValue;
  final double maxValue;
  final bool isVisible;

  AutomationLane({
    required this.id,
    required this.parameterId,
    required this.type,
    List<AutomationPoint>? points,
    Color? curveColor,
    double? minValue,
    double? maxValue,
    this.isVisible = true,
  })  : points = points ?? [],
        curveColor = curveColor ?? _defaultColorForType(type),
        minValue = minValue ?? _defaultMinForType(type),
        maxValue = maxValue ?? _defaultMaxForType(type);

  /// Get interpolated value at specific time
  double getValueAt(double timeSeconds) {
    if (points.isEmpty) return minValue;
    if (points.length == 1) return _denormalize(points[0].value);

    // Sort points by time
    final sortedPoints = List<AutomationPoint>.from(points)
      ..sort((a, b) => a.time.compareTo(b.time));

    // Find surrounding points
    AutomationPoint? before;
    AutomationPoint? after;

    for (int i = 0; i < sortedPoints.length; i++) {
      final point = sortedPoints[i];
      if (point.time <= timeSeconds) {
        before = point;
      }
      if (point.time >= timeSeconds && after == null) {
        after = point;
        break;
      }
    }

    // Edge cases
    if (before == null) return _denormalize(sortedPoints.first.value);
    if (after == null) return _denormalize(sortedPoints.last.value);
    if (before.time == after.time) return _denormalize(before.value);

    // Interpolate based on curve type
    final t = (timeSeconds - before.time) / (after.time - before.time);
    final interpolatedNormalized = _interpolate(before, after, t);

    return _denormalize(interpolatedNormalized);
  }

  /// Interpolate between two points
  double _interpolate(AutomationPoint a, AutomationPoint b, double t) {
    switch (a.interpolation) {
      case CurveType.step:
        return a.value;

      case CurveType.linear:
        return a.value + (b.value - a.value) * t;

      case CurveType.bezier:
        // Simple bezier: control point at midpoint
        final cp = (a.value + b.value) / 2;
        final u = 1 - t;
        return u * u * a.value + 2 * u * t * cp + t * t * b.value;

      case CurveType.exponential:
        return a.value + (b.value - a.value) * (t * t);

      case CurveType.logarithmic:
        return a.value + (b.value - a.value) * math.sqrt(t);
    }
  }

  /// Convert normalized 0-1 to actual parameter value
  double _denormalize(double normalized) {
    return minValue + (maxValue - minValue) * normalized.clamp(0.0, 1.0);
  }

  /// Convert actual parameter value to normalized 0-1
  double normalize(double value) {
    if (maxValue == minValue) return 0.0;
    return ((value - minValue) / (maxValue - minValue)).clamp(0.0, 1.0);
  }

  /// Add automation point
  AutomationLane addPoint(AutomationPoint point) {
    final updatedPoints = List<AutomationPoint>.from(points)..add(point);
    return copyWith(points: updatedPoints);
  }

  /// Remove point by ID
  AutomationLane removePoint(String pointId) {
    final updatedPoints = points.where((p) => p.id != pointId).toList();
    return copyWith(points: updatedPoints);
  }

  /// Update existing point
  AutomationLane updatePoint(String pointId, {double? time, double? value, CurveType? interpolation}) {
    final updatedPoints = points.map((p) {
      if (p.id == pointId) {
        return p.copyWith(time: time, value: value, interpolation: interpolation);
      }
      return p;
    }).toList();
    return copyWith(points: updatedPoints);
  }

  /// Smooth curve between two points (add intermediate points)
  AutomationLane smoothBetween(int startIndex, int endIndex) {
    if (startIndex >= endIndex || startIndex < 0 || endIndex >= points.length) {
      return this;
    }

    final sortedPoints = List<AutomationPoint>.from(points)
      ..sort((a, b) => a.time.compareTo(b.time));

    final start = sortedPoints[startIndex];
    final end = sortedPoints[endIndex];

    // Add 3 intermediate points with bezier curve
    final newPoints = List<AutomationPoint>.from(sortedPoints);
    final timeDelta = (end.time - start.time) / 4;

    for (int i = 1; i <= 3; i++) {
      final t = i / 4.0;
      final time = start.time + timeDelta * i;
      final value = _interpolate(start, end, t);

      newPoints.add(AutomationPoint(
        id: 'smooth_${start.id}_${end.id}_$i',
        time: time,
        value: value,
        interpolation: CurveType.bezier,
      ));
    }

    return copyWith(points: newPoints);
  }

  /// Copy with modifications
  AutomationLane copyWith({
    String? id,
    String? parameterId,
    AutomationParameterType? type,
    List<AutomationPoint>? points,
    Color? curveColor,
    double? minValue,
    double? maxValue,
    bool? isVisible,
  }) {
    return AutomationLane(
      id: id ?? this.id,
      parameterId: parameterId ?? this.parameterId,
      type: type ?? this.type,
      points: points ?? this.points,
      curveColor: curveColor ?? this.curveColor,
      minValue: minValue ?? this.minValue,
      maxValue: maxValue ?? this.maxValue,
      isVisible: isVisible ?? this.isVisible,
    );
  }

  /// JSON serialization
  Map<String, dynamic> toJson() => {
    'id': id,
    'parameterId': parameterId,
    'type': type.name,
    'points': points.map((p) => p.toJson()).toList(),
    'curveColor': curveColor.value,
    'minValue': minValue,
    'maxValue': maxValue,
    'isVisible': isVisible,
  };

  factory AutomationLane.fromJson(Map<String, dynamic> json) {
    return AutomationLane(
      id: json['id'] as String,
      parameterId: json['parameterId'] as String,
      type: AutomationParameterType.values.firstWhere((t) => t.name == json['type']),
      points: (json['points'] as List).map((p) => AutomationPoint.fromJson(p)).toList(),
      curveColor: Color(json['curveColor'] as int),
      minValue: (json['minValue'] as num).toDouble(),
      maxValue: (json['maxValue'] as num).toDouble(),
      isVisible: json['isVisible'] as bool? ?? true,
    );
  }

  /// Default color for parameter type
  static Color _defaultColorForType(AutomationParameterType type) {
    switch (type) {
      case AutomationParameterType.volume:
        return const Color(0xFFFF9040); // Orange
      case AutomationParameterType.pan:
        return const Color(0xFF40C8FF); // Cyan
      case AutomationParameterType.rtpc:
        return const Color(0xFF9370DB); // Purple
      case AutomationParameterType.trigger:
        return const Color(0xFF40FF90); // Green
    }
  }

  /// Default min value for parameter type
  static double _defaultMinForType(AutomationParameterType type) {
    switch (type) {
      case AutomationParameterType.volume:
        return 0.0;
      case AutomationParameterType.pan:
        return -1.0;
      case AutomationParameterType.rtpc:
        return 0.0;
      case AutomationParameterType.trigger:
        return 0.0;
    }
  }

  /// Default max value for parameter type
  static double _defaultMaxForType(AutomationParameterType type) {
    switch (type) {
      case AutomationParameterType.volume:
        return 2.0; // +6dB
      case AutomationParameterType.pan:
        return 1.0;
      case AutomationParameterType.rtpc:
        return 1.0;
      case AutomationParameterType.trigger:
        return 1.0;
    }
  }
}
